### PR TITLE
Slice buffer reallocate design::PR02::refactoring for some basic function

### PR DESCRIPTION
**Pull request by shihuade**
--refactoring for RcCalculateGomQp etc. functions
--refactoring for WlesMarkMMCORefInfo
--refactoring for InitSliceHeadWithBase()
--refactoring for WelsGetNumMbInSlice

code review:
[1687](https://rbcommons.com/s/OpenH264/r/1687/)
[1688](https://rbcommons.com/s/OpenH264/r/1688/)
[1689](https://rbcommons.com/s/OpenH264/r/1689/)
[1691](https://rbcommons.com/s/OpenH264/r/1691/)

### CODE REVIEW

File: codec/encoder/core/src/encoder_ext.cpp [hunk 1]
```diff
@@ -3787,6 +3780,18 @@ int32_t WelsEncoderEncodeExt (sWelsEncCtx* pCtx, SFrameBSInfo* pFbi, const SSour
         //TO DO: add update ppSliceInLayer module based on pSliceInThread[ThreadNum]
         // UpdateSliceInLayerInfo(); // reordering
 
+        //TO DO: add update ppSliceInLayer module based on pSliceInThread[ThreadNum]
```
Comment (sijchen) [inline codec/encoder/core/src/encoder_ext.cpp]: why there are redundant comments here?

File: codec/encoder/core/src/svc_encode_slice.cpp [hunk 2]
```diff
@@ -1197,11 +1200,8 @@ int32_t ReallocateSliceList (sWelsEncCtx* pCtx,
       return iRet;
     }
 
-    iRet = InitSliceHeadWithBase (pSlice, pBaseSlice, pCtx->iNumRef0);
-    if (ENC_RETURN_SUCCESS != iRet) {
-      FreeSliceBuffer(pNewSliceList, kiMaxSliceNumNew, pMA, "ReallocateSliceList()::InitSliceBsBuffer()");
-      return iRet;
-    }
+    InitSliceHeadWithBase (pSlice, pBaseSlice);
```
Comment (sijchen) [inline codec/encoder/core/src/svc_encode_slice.cpp]: as same protection as previous codes, pls add at line 1167 such like:
     if (NULL == pBaseSlice) {
       FreeSliceBuffer(pNewSliceList, kiMaxSliceNumNew, pMA, "ReallocateSliceList()::InitSliceBsBuffer()");		       FreeSliceBuffer(pNewSliceList, kiMaxSliceNumNew, pMA, "ReallocateSliceList()::InitSliceBsBuffer()");
       return ENC_RETURN_MEMALLOCERR;		       return ENC_RETURN_MEMALLOCERR;
     }		     }

File: codec/encoder/core/src/encoder_ext.cpp (hunk 1 above)
Comment (shihuade) [inline codec/encoder/core/src/encoder_ext.cpp]: oh, seems it produced by cherry-pick command, has been remove

File: codec/encoder/core/src/svc_encode_slice.cpp (hunk 2 above)
Comment (shihuade) [inline codec/encoder/core/src/svc_encode_slice.cpp]: fixed.

**Comment by sijchen** [comment]
Archiving from https://rbcommons.com/s/OpenH264/r/1687/

----
**testing_done**:  
**description**:  <p>refactor for InitSliceHeadWithBase()<br />1. break InitSliceHeadWithBase() into tow small functions<br />  InitSliceHeadWithBase()<br />  InitSliceRefInfoWithBase（）</p><p>so InitSliceHeadWithBase() <br />    can be reused in <br />    WelsInitCurrentLayer()</p><ol start="2" style="counter-reset: li 1"><li>change to return type from int to void.<br />   there is check before calling these functions</li></ol>

----
 reviewer Karina Li (ruil2) 
 time 2016-09-09T09:39:58+00:00 
2016-09-09T09:39:58+00:00 Karina Li (ruil2):
 on **codec/encoder/core/src/svc_encode_slice.cpp line1105**: 
 id="comment447623"><ol><li><p>InitSliceHeadWithBase is enough, there is no neccessary to add a new one.</p></li><li><p>are there any possible that pSlice or  pBaseSlice is  NULL?</p></li></ol>
InitSliceHeadWithBase is enough, there is no neccessary to add a new one.
are there any possible that pSlice or  pBaseSlice is  NULL?

2016-09-30T05:23:47+00:00 Forrest Shi (huade):
><ol><li>reduce code, InitSliceHeadWithBase()  can be reused in  WelsInitCurrentLayer()<br />2.will not call thest two function if pSlice or pBaseSlice is NULL</li></ol>


----
 reviewer Karina Li (ruil2) 
 time 2016-10-10T02:26:19+00:00 
2016-10-10T02:26:19+00:00 Karina Li (ruil2):
>Ship It!
Ship It!
Ship it!


----

----

Finish archiving from https://rbcommons.com/s/OpenH264/r/1687/

**Comment by sijchen** [comment]
Archiving from https://rbcommons.com/s/OpenH264/r/1688/

----
**testing_done**:  
**description**:  <p>refactor for WlesMarkMMCORefInfo<br />1. common part of WlesMarkMMCORefInfo()<br />    and <br />   WlesMarkMMCORefInfoScreen()<br />  to be new function <br />      WlesMarkMMCORefInfoWithBase() <br />2. instead of init slice by slice,<br />   in new design, init with pBaseSlice.</p>

----
 reviewer Karina Li (ruil2) 
 time 2016-09-13T03:56:42+00:00 
2016-09-13T03:56:42+00:00 Karina Li (ruil2):
 on **codec/encoder/core/src/ref_list_mgr_svc.cpp line460**: 
 id="comment448677"><ol><li>slice 0 had been initialized before the function calling.</li><li>what merit on this refactor? In fact,each slice can have the different header info. this change limits the extension.</li></ol>

2016-09-30T05:35:27+00:00 Forrest Shi (huade):
><h1>1, no, sliceIndex=0 not been init, it is inited the as other sliceindex=1,2 ...</h1><div class="codehilite"><pre><span></span>you can refer to for loopfor (int32_t iSliceIdx = 0; iSliceIdx &lt; kiCountSliceNum; iSliceIdx++) {...}
this refoctor, the purpose is WlesMarkMMCORefInfoWithBase() can be reused both in <br />      WlesMarkMMCORefInfo() and WlesMarkMMCORefInfoScreen()<br />  and there is no logic change


----
 reviewer Karina Li (ruil2) 
 time 2016-10-10T02:26:07+00:00 
2016-10-10T02:26:07+00:00 Karina Li (ruil2):
>Ship It!
Ship It!
Ship it!


----

----

Finish archiving from https://rbcommons.com/s/OpenH264/r/1688/

**Comment by sijchen** [comment]
Archiving from https://rbcommons.com/s/OpenH264/r/1689/

----
**testing_done**:  
**description**:  <p>refactoring for RcCalculateGomQp etc. functions<br /> pSlice <br />    instead of <br /> iSliceIndex <br />   as input parameter in below functions:<br />    RcCalculateMbQp()<br />    RcGomTargetBits()<br />    RcCalculateGomQp()</p>

----
 reviewer Karina Li (ruil2) 
 time 2016-11-02T08:51:24+00:00 
2016-11-02T08:51:24+00:00 Karina Li (ruil2):
>Ship It!
Ship It!
Ship it!


----

----

Finish archiving from https://rbcommons.com/s/OpenH264/r/1689/

**Comment by sijchen** [comment]
Archiving from https://rbcommons.com/s/OpenH264/r/1691/

----
**testing_done**:  
**description**:  <ol><li><p>refactoring for WelsGetNumMbInSlice()<br />   ----change validate check logic<br />   ----add pSlice as input parameters<br />       so not limit to pSlice from<br />           --pCurDq-&gt;ppSliceInLayer[kuiSliceIdc]<br />       but also for other cases like:<br />           --pCurLayer-&gt;sSliceThreadInfo.pSliceInThread[iThreadIdx] + iSliceIdx</p></li><li><p>remove ppSliceList from below funtion:<br />   WelsMarkPic()<br />   WelsUpdateRefSyntax()<br />   UpdateOriginalPicInfo()</p></li></ol>

----
 reviewer Karina Li (ruil2) 
 time 2016-11-02T08:58:24+00:00 
2016-11-02T08:58:24+00:00 Karina Li (ruil2):
>Ship It!
Ship It!
Ship it!


----

----

Finish archiving from https://rbcommons.com/s/OpenH264/r/1691/

**Pull request merged**